### PR TITLE
Serialize witness

### PR DESCRIPTION
**Pull request by xgaozoyoe**
1. Support load and store witness [done]
2. Simple tests [done]

### CODE REVIEW

File: halo2_proofs/src/helpers.rs [hunk 1]
```diff
@@ -14,3 +28,840 @@ pub(crate) trait CurveRead: CurveAffine {
...
+impl<T: Serializable> Serializable for (T, T) {
+    fn fetch<R: io::Read>(reader: &mut R) -> io::Result<Self> {
+        Ok((T::fetch(reader)?, T::fetch(reader)?))
+    }
+    fn store<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.0.store(writer)?;
+        self.1.store(writer)?;
+        Ok(())
+    }
+
```
Comment (junyu0312) [inline halo2_proofs/src/helpers.rs:94]: format

File: halo2_proofs/src/helpers.rs [hunk 2]
```diff
@@ -14,3 +28,840 @@ pub(crate) trait CurveRead: CurveAffine {
...
+    fn store<W: io::Write>(&self, writer: &mut W) -> io::Result<()>;
+}
+
+fn read_u32<R: io::Read>(reader: &mut R) -> io::Result<u32> {
+    let mut r = [0u8; 4];
+    reader.read(&mut r)?;
+    Ok(u32::from_le_bytes(r))
+}
+
+impl Serializable for usize {
```
Comment (junyu0312) [inline halo2_proofs/src/helpers.rs]: impl Serializable for u32, since `usize` depends on architecture

File: halo2_proofs/src/helpers.rs [hunk 3]
```diff
@@ -14,3 +28,840 @@ pub(crate) trait CurveRead: CurveAffine {
...
+}
+
+impl<C: CurveAffine> Serializable for VerifyingKey<C> {
+    fn store<W: io::Write>(
+        &self,
+        writer: &mut W,
+    ) -> io::Result<()> {
+        let j = (self.domain.get_quotient_poly_degree() + 1) as u32; // quotient_poly_degree is j-1
+        let k = self.domain.k() as u32;
+        println!("store vkey j is {}, k is {}", j ,k);
```
Comment (junyu0312) [inline halo2_proofs/src/helpers.rs]: remove

**Review by junyu0312 — APPROVED** [review]

File: halo2_proofs/src/plonk/keygen.rs [hunk 4]
```diff
@@ -349,3 +349,118 @@ where
...
+    params: &Params<C>,
+    vk: &VerifyingKey<C>,
+    fixed: Vec<Polynomial<C::Scalar, LagrangeCoeff>>,
+    permutation: permutation::keygen::Assembly,
+) -> Result<ProvingKey<C>, Error>
+where
+    C: CurveAffine,
+{
+    let cs = vk.cs.clone();
+    let selectors = vec![vec![false; params.n as usize]; cs.num_selectors];
```
Comment (lanbones) [inline halo2_proofs/src/plonk/keygen.rs]: We'd better to assert selector number == 0. It cannot handle selector correctly.

File: halo2_proofs/src/plonk/circuit.rs [hunk 5]
```diff
@@ -1469,6 +1471,18 @@ impl<F: Field> ConstraintSystem<F> {
         tmp
     }
 
+    /// Allocate a new advice column
+    pub fn named_advice_column(&mut self, name: String) -> Column<Advice> {
+        let tmp = Column {
```
Comment (lanbones) [inline halo2_proofs/src/plonk/circuit.rs]: rename to `res`.

File: halo2_proofs/src/plonk/keygen.rs [hunk 6]
```diff
@@ -349,3 +349,118 @@ where
...
+        ev,
+    })
+}
+
+/// Generate a `ProvingKey` from a `VerifyingKey` and an instance of `Circuit`.
+pub(crate) fn generate_pk_info<C, ConcreteCircuit>(
+    params: &Params<C>,
+    vk: &VerifyingKey<C>,
+    circuit: &ConcreteCircuit,
+) -> Result<(Vec<Polynomial<C::Scalar, LagrangeCoeff>>,permutation::keygen::Assembly), Error>
```
Comment (lanbones) [inline halo2_proofs/src/plonk/keygen.rs]: format

File: halo2_proofs/src/plonk/keygen.rs [hunk 7]
```diff
@@ -349,3 +349,118 @@ where
...
+where
+    C: CurveAffine,
+    ConcreteCircuit: Circuit<C::Scalar>,
+{
+    let mut cs = ConstraintSystem::default();
+    let config = ConcreteCircuit::configure(&mut cs);
+
+    let cs = cs;
```
Comment (lanbones) [inline halo2_proofs/src/plonk/keygen.rs]: remove

File: halo2_proofs/src/plonk/prover.rs [hunk 8]
```diff
@@ -294,6 +296,457 @@ pub fn create_proof<
...
+
+    let advice: Vec<AdviceSingle<C>> =
+        instances.iter()
+        .map(|_| -> Result<AdviceSingle<C>, Error> {
+            let unusable_rows_start = params.n as usize - (meta.blinding_factors() + 1);
+
+            let mut advice = AssignWitnessCollection::fetch_witness(params, fd)?;
+
             // Add blinding factors to advice columns
             for advice in &mut advice {
```
Comment (lanbones) [inline halo2_proofs/src/plonk/prover.rs]: need to skip adding blind value for named column here

**Review by lanbones — APPROVED** [review]

**Pull request merged**